Try: Overlay caption w. text-shadow.

jasmussen opened the pull request:
## What?

Try to make progress on #8030 and #56587 by improving the caption style.

The current dark scrim is a very dark gradient to aid legibility for the white text on light backgrounds, however the sharpness of the gradient itself makes the legibility sub-optimal:

![overlay-caption-before](https://github.com/user-attachments/assets/258e59c6-5836-438b-bd2f-601eaaa127de)

This PR does a few things to keep the spirit of the same style, but improves it by:

* Adding a small text-shadow to the text itself.
* Applying a backdrop-blur to allow the text silhouette to avoid conflicts with complex imagery below.
* Changes the padding to be em based, and slightly more relaxed.
* Changes the color stops of the gradients to afford better contrast not just when it's one line of text, but multiple lines as well.

![overlay-caption-after](https://github.com/user-attachments/assets/82963571-a8af-43cc-b685-490e4976e1bf)

## Testing Instructions

Insert galleries with overlay captions.

jasmussen committed:
Try: Overlay caption w. text-shadow.

beafialho commented:
Nice! This may be perhaps out of scope but has the option of a left/right aligned overlay caption been considered?

jasmussen commented:
I'd love for a separate "Caption" element to be stylable, just like buttons, forms, inputs, etc. Right now that's missing, and I'd like to see that, this would also allow font size changes, alignments, color changes, etc.

jasmussen committed:
Tweak text-shadow.

jasmussen commented:
Marking this one ready for review. The combination of background blur, a refined gradient, and the text shadow keeps contrast the same as before in most contexts, but improves it when many lines of overlay caption text is used. It's also close enough to the existing style that it is firmly in improvement territory, which is important given this is already shipping and in use. Would appreciate your reviews!

t-hamano reviewed (commented):
This is a great improvement. However, what I'm seeing seems a little different for some reason.

There's space on the left and right of the scrim. After some research, it seems like this is being affected by `scrollbar-gutter: stable both-edges;` which is part of the style of the `custom-scrollbars-on-hover` mixin. This issue only appears in certain environments (Windows OS, Chrome):

| Chrome | Firefox |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/280e4e10-d3d6-4fff-b3de-6947d4329e15) | ![image](https://github.com/user-attachments/assets/84ba2031-0a53-4fc6-83e2-c31eaa4ca687)| 

I also believe that we need to address two issues.

If the caption is scrollable, the bottom position of the scrim will move:

![image](https://github.com/user-attachments/assets/89bed94f-0245-48ea-851a-93b6fd3356bf)

This style needs to be removed when the Image block is in the "Rounded" style:

![image](https://github.com/user-attachments/assets/1864814a-7b3e-4df7-b111-dff839a28b04)

jasmussen commented:
Just wanted to note that I really appreciate your excellent feedback, and will address it! Other things got in the way first, but the review hasn't been forgotten. Thank you 🙏

jasmussen committed:
Address feedback.

jasmussen committed:
Use "before" instead of "after".

paaljoachim commented:
Btw
Here is the issue related to Image caption alignments:
https://github.com/WordPress/gutenberg/issues/12997

jasmussen committed:
Address issues with external captions.

jasmussen commented:
Took a stab. This was harder to address than I thought it would've been, so it needed kind of a fresh approach. What I've done now, and I believe this actually addresses all the items you brought up (though good to test again), is this:

* I kept the adjusted/tweaked dark scrim gradient on the `figcaption` element itself. This addresses issues of the scrim only ever being as tall as the caption.
* I've moved the _blur_ layer to be attached to the `figure` itself, but only when there's a figcaption. This uses `:has()`, which has support, but that entire layer will simply disappear for IE11, and is fine. 

The combination of these two, I feel, still captures the spirit of the effort, but should work better:

![Screenshot 2024-08-13 at 16 54 36](https://github.com/user-attachments/assets/3c243093-14fe-461e-ab2b-1cfbf05d34a5)

t-hamano commented on packages/block-library/src/gallery/style.scss:
```diff
@@ -43,22 +43,46 @@ figure.wp-block-gallery.has-nested-images {
+		// Create a background blur layer.
+		&:has(figcaption)::before {
+			content: "";
+			z-index: 0;
```
There doesn't seem to be any visual difference without `z-index: 0`, but is it necessary?

t-hamano commented on packages/block-library/src/gallery/style.scss:
```diff
@@ -43,22 +43,46 @@ figure.wp-block-gallery.has-nested-images {
+
+		// Place the caption at the bottom.
+		figcaption {
+			z-index: 1;
```
There doesn't seem to be any visual difference without `z-index: 1`, but is it necessary?

jasmussen replied:
Good catch on both z-indexes. They were briefly necessary in one of the versions I was testing, but in this final version they are not.

t-hamano commented on packages/block-library/src/gallery/style.scss:
```diff
@@ -43,22 +43,46 @@ figure.wp-block-gallery.has-nested-images {
+		figcaption {
+			z-index: 1;
 			color: $white;
+			text-shadow: 0 0 1.5px rgba($color: $black, $alpha: 1);
```
```suggestion
			text-shadow: 0 0 1.5px $black;
```

Since `rgba($color: $black, $alpha: 1)` is the same as black, I think it can be made simpler.

t-hamano commented on packages/block-library/src/gallery/style.scss:
```diff
@@ -43,22 +43,46 @@ figure.wp-block-gallery.has-nested-images {
 			@include custom-scrollbars-on-hover(transparent, rgba($white, 0.8));
 
+			// Dark gradient scrim.
+			background: linear-gradient(0deg, rgba($color: $black, $alpha: 0.4) 0%, rgba($color: $black, $alpha: 0) 100%, transparent);
```
```suggestion
			background: linear-gradient(0deg, rgba($color: $black, $alpha: 0.4) 0%, transparent 100%);
```

The gradient style should remain the same.

t-hamano reviewed (commented):
Thanks for the update! It looks like it's almost complete.

After testing it again, I noticed the following.

Personally, I feel that `blur(4px)` has a slightly large impact depending on the content of the image. What do you think about 2px?

4px blur:

![image](https://github.com/user-attachments/assets/bb691751-ed57-4262-a5dd-f95c98014311)

2px blur:

![image](https://github.com/user-attachments/assets/c61d53bd-54b2-44d2-a009-59b350fb013e)

Another thing I noticed is that the caption height is not constrained. Should we apply `max-height: 40%` if we want it to match the scrim?

![image](https://github.com/user-attachments/assets/0f592252-ee5d-4490-8161-332cb96bbdf2)

jasmussen commented:
Great feedback as always, thank you.

> Personally, I feel that blur(4px) has a slightly large impact depending on the content of the image. What do you think about 2px?

Can we meet at 3px?

![Screenshot 2024-08-16 at 09 42 05](https://github.com/user-attachments/assets/de3b0b26-6720-44a4-9b0d-9adfb7154ef0)

2px feels so small to me that it might as well not be there:

![Screenshot 2024-08-16 at 09 41 56](https://github.com/user-attachments/assets/dc502454-3264-4ad1-bdfe-9a9803bcddd7)

And by that I mean, if we're going to do it, it should be perceptible, otherwise best to not do it. 

I'll go through your other pieces one by one now!

jasmussen committed:
Address feedback.

jasmussen commented:
> Another thing I noticed is that the caption height is not constrained. Should we apply max-height: 40% if we want it to match the scrim?

I intentionally let it be unconstrained, what I found as I tinkered was that when constrained like we do now, the scrollbar behaves somewhat curiously since it only scrolls the caption. That also means that the caption itself can be cropped, which can look awkward, like so:

![Screenshot 2024-08-16 at 09 47 08](https://github.com/user-attachments/assets/2c7ae6a3-a0f6-46cd-a0fb-c1c7683cb9a8)

Finally for really small images with captions, if we constrain the height, there might not be space to show anything at all. 

That said, I'm happy to restore it to keep this PR smaller: it's mainly about elevating the quality and clarity of the scrim, than changing the behavior, so no strong opinion from my end. Let me know if I should restore it!

t-hamano commented on packages/block-library/src/gallery/style.scss:
```diff
@@ -43,22 +43,45 @@ figure.wp-block-gallery.has-nested-images {
+			// Blur the background under the gradient scrim.
+			backdrop-filter: blur(3px);
+
+			// Crop the caption so the blur is only on the edge.
+			mask-image: linear-gradient(0deg, rgba($color: $black, $alpha: 1) 20%, rgba($color: $black, $alpha: 0) 100%);
```
```suggestion
			// Crop the caption so the blur is only on the edge.
			mask-image: linear-gradient(0deg, $black 20%, transparent 100%);
```

t-hamano commented on packages/block-library/src/gallery/style.scss:
```diff
@@ -43,22 +43,45 @@ figure.wp-block-gallery.has-nested-images {
 			box-sizing: border-box;
 			@include custom-scrollbars-on-hover(transparent, rgba($white, 0.8));
 
+			// Dark gradient scrim.
+			//background: linear-gradient(0deg, rgba($color: $black, $alpha: 0.4) 0%, rgba($color: $black, $alpha: 0) 100%, transparent);
+			background: linear-gradient(0deg, rgba($color: $black, $alpha: 0.4) 0%, transparent 100%);
```
```suggestion
			// Dark gradient scrim.
			background: linear-gradient(0deg, rgba($color: $black, $alpha: 0.4) 0%, transparent 100%);
```

t-hamano commented on packages/block-library/src/gallery/style.scss:
```diff
@@ -43,22 +43,45 @@ figure.wp-block-gallery.has-nested-images {
 			overflow: auto;
-			padding: 0 8px 8px;
-			position: absolute;
+			padding: 1em;
```
This is a non-blocker suggestion.

Because `scrollbar-gutter: stable both-edges` has been applied to the `figcaption`, the padding on the left and right looks wider, we may want to make it a little smaller.

Below is what it looks like on my environment, but how does it look on yours?

`padding: 1em`:

![image](https://github.com/user-attachments/assets/d3660bc1-261d-42b8-a77a-5a1258fc7909)

`padding: 1em 0.5em`:

![image](https://github.com/user-attachments/assets/b512376a-3c75-4609-b647-ae07d8fa8d49)

jasmussen replied:
I don't disagree, though since this PR already touches a fair bit of CSS, I wonder if it's not best to do this separately?

Thank you again for the thorough reviews!

t-hamano reviewed (approved):
LGTM! I left just a few suggestions at the end, but it all seems to work fine.

> Finally for really small images with captions, if we constrain the height, there might not be space to show anything at all.

I had overlooked this scenario. It certainly seems like it would be better to leave the height unconstrained.

I was concerned that the lightbox toggle button would become unclickable when the caption reached 100% height, but that doesn't seem to be an issue:

![image](https://github.com/user-attachments/assets/d77d094c-ca49-4525-b3fb-4a873a85426f)

jasmussen committed:
Update packages/block-library/src/gallery/style.scss

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

jasmussen committed:
Update packages/block-library/src/gallery/style.scss

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

jasmussen commented:
Just checking, is this good to merge?

jasmussen commented:
I'm going to merge, but happy to follow up!

t-hamano commented:
Sorry for the late reply. Of course, my suggestion regarding padding could be tried in another PR 👍

paaljoachim commented:
Hei Joen.

I have a comment here...

In Gutenberg plugin 19.1 Gallery Image captions now look like this:

![Screenshot 2024-09-03 at 21 30 59](https://github.com/user-attachments/assets/bbdc7b52-4369-4d80-bd3c-0e49312555c3)

The artist web site I am creating the artist does not want the caption on the images inside galleries so I moved these below with the following code.

```
.wp-block-gallery.has-nested-images figure.wp-block-image figcaption {
    background: 0 0;
    color: #C0BBBB;	
    position: initial;
    text-align: left;
}
```

The result looks like this:
![Screenshot 2024-09-03 at 21 33 02](https://github.com/user-attachments/assets/13d8f5fb-f8ad-4e2b-962c-95e8f0a1bff6)

I should have tested this PR before it was merged with the setup I have for the web site, but I did not think about it.
This means when the caption is on the images the text looks nice and sharp if the caption is moved below the image it becomes blurry.

I am now having a difficult time using an important to remove the text shadow in the custom code.
`text-shadow: 0 0 1.5px #000;`

paaljoachim commented:
One more thing. I noticed the bottom part of the Gallery images are blurred where the caption was earlier. 

Gutenberg 19.1 active.
![Screenshot 2024-09-03 at 21 52 56](https://github.com/user-attachments/assets/2198b6f4-abe3-498e-9ebd-d5ce7abdbb5a)

Gutenberg 19.1 not active.
![Screenshot 2024-09-03 at 21 52 37](https://github.com/user-attachments/assets/28fc1505-e8e8-45fc-9cd4-43475d3e6eca)

jasmussen commented:
Can you try adding this CSS to your custom override?

```
.wp-block-gallery.has-nested-images figure.wp-block-image:has(figcaption):before {
	-webkit-backdrop-filter: none;
	backdrop-filter: none;
}
```

That said, it's not clear to me this is necessarily an issue with the CSS as written, since it's meant to be an overlay.

One thing we found out is that, although there's no official support for captions below the image in a gallery, there's a little trick. If you apply a border-radius to the image in the gallery, the caption gets pushed below:
![2024-09-04 09 36 04 tt5 local e451ed5fea24](https://github.com/user-attachments/assets/0f6f6ec1-d2cb-4281-8a7d-3135a5c4b7f6)

CC: @aaronrobertshaw I noticed that this trick to move the caption below the image only works when the radius is applied locally. If you apply a border radius to the Image block in global styles, it does not change the galleries. Could it? Should it? It seems like if the behavior is in place locally, it should be in place globally as well, no?

paaljoachim commented:
```
.wp-block-gallery.has-nested-images figure.wp-block-image:has(figcaption):before {
	-webkit-backdrop-filter: none;
	backdrop-filter: none;
}
```

Thank you Joen!
The above code removes the blurred overlay in the bottom of the images inside a gallery where the caption it moved below the image.

The ideal would be go with....

> Whether captions should be overlayed or not seems to me like it should be a block setting or perhaps a block style variation. Better yet, I'm still a fan of a https://github.com/WordPress/gutenberg/pull/31823 which could encapsulate this sort of behaviour and styling.

As @aaronrobertshaw mentioned. As it would hopefully make it openly possible for a user to choose the location (and have editing tools) of the caption in relation to the image in an Image block and a Gallery block.
